core/modules/qdisp/Executive.cc

```diff
@@ -150,61 +119,52 @@ Executive::Executive(Config::Ptr c, std::shared_ptr<MessageStore> ms)
-        // If the reset fails, do nothing-- can't retry.
-    }
-private:
-    Executive& _executive;
-    int _jobId;
-    JobDescription _jobDescription;
-    JobStatus::Ptr _jobStatus; ///< Points at status in Executive::_statusMap
-};
-
-/* Add a new job to executive queue, if not already in. Not thread-safe.
+/** Add a new job to executive queue, if not already in. Not thread-safe.
  */
-void Executive::add(int jobId, Executive::JobDescription const& jobDescription) {
-    bool alreadyCancelled = lockedRead(_cancelled, _cancelledMutex);
-    if(alreadyCancelled) {
-        LOGF(getLogger(), LOG_LVL_INFO, "Executive already cancelled, ignoring add(%1%)" % jobId);
+void Executive::add(JobDescription const& jobDesc) {
+    LOGF_DEBUG("Executive::add(%1%)" % jobDesc.toString());
+    if(_cancelled.get()) {
+        LOGF(getLogger(), LOG_LVL_INFO, "Executive already cancelled, ignoring add(%1%)" % jobDesc.id());
         return;
     }
-    bool trackOk = _track(jobId, jobDescription.requester); // Remember so we can join
-    if(!trackOk) {
-        LOGF(getLogger(), LOG_LVL_WARN, "Ignoring duplicate add(%1%)" % jobId);
+    // Create the JobQuery and put it in the map.
+    JobStatus::Ptr jobStatus = std::make_shared<JobStatus>();
+    MarkCompleteFunc::Ptr mcf = std::make_shared<MarkCompleteFunc>(this, jobDesc.id());
+    JobQuery::Ptr jobQuery(new JobQuery(this, jobDesc, jobStatus, mcf));
+    jobQuery->setupWeakThis(jobQuery);
+    if(!_addJobToMap(jobQuery)) {
+        LOGF_ERROR("Executive ignoring duplicate job add(%1%)" % jobQuery->getId());
         return;
     }
-    JobStatus::Ptr jobStatus = _insertNewStatus(jobId, jobDescription.resource);
 
-    if (_empty.get()) {
-        _empty.set(false);
-        LOGF(getLogger(), LOG_LVL_TRACE, "Flag \"Empty\" set to false by jobId %1%" % jobId);
+    if (!_track(jobQuery->getId(), jobQuery)) {
+        LOGF_ERROR("Executive ignoring duplicate track add(%1%)" % jobQuery->getId());
+        return;
+    }
+    if (_empty.set(false)) {
+        LOGF_INFO("Flag _empty set to false by jobId %1%" % jobQuery->getId());
     }
-
     ++_requestCount;
-    std::string msg = "Executive: Add job with path=" + jobDescription.resource.path();
-    LOGF(getLogger(), LOG_LVL_INFO, "%1%" % msg);
-    _messageStore->addMessage(jobDescription.resource.chunk(), log::MSG_MGR_ADD, msg);
+    std::string msg = "Executive: Add job with path=" + jobDesc.resource().path();
+    LOGF_INFO("%1%" % msg);
+    _messageStore->addMessage(jobDesc.resource().chunk(), log::MSG_MGR_ADD, msg);
+
+    jobQuery->runJob();
+}
 
-    _dispatchQuery(jobId,
-                   jobDescription,
-                   jobStatus);
+/** Add a JobQuery to this Executive.
+ * Return true if it was successfully added to the map.
+ */
+bool Executive::_addJobToMap(JobQuery::Ptr job) {
+    int jobId = job->getId();
+    if (_jobMap.find(jobId) != _jobMap.end()) {
+        return false;
+     }
+     _jobMap[jobId] = job;
+     return true;
 }
```

Comment:
This whole function is equivalent to:

    return _jobMap.insert(job->getId(), job).second;

(except that it is less efficient as it needs 2 lookups).